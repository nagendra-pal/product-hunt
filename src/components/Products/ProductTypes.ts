// All types related to product defined here

export interface ProductType {
    id: number;
    name: string;
    image: string;
    desc: string;
    likes: number;
    comments: {text: string}[];
    created: Date;
};

export interface DateType {
    startDate: Date;
    endDate: Date;
};

export interface ProductListType {
    productList: Array<ProductType>;
};

export interface ProductListItemType {
    product: ProductType;
};
