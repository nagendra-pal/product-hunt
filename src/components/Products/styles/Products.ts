
import styled from "styled-components";
import { Button } from "reactstrap";

export const Body = styled.div`
  padding: 1rem;
`;

export const Title = styled.div`
  font-size: 2rem;
  border-bottom: 0.0625rem solid #e8e8e8;
`;

export const SearchContainer = styled.div`
  display: flex;
  flex-direction: column;
  padding: 1.5rem 1.875rem 2rem;
  margin: 0 -1.875rem 1.3125rem;
  border-bottom: 0.0625rem solid #e8e8e8;
  @media only screen and (min-width: 768px) {
    flex-direction: row;
    padding: 1.5rem 0 2rem;
    margin: 0 0 1.3125rem;
  }
`;

export const DateSearchLabel = styled.div`
  font-size: 1rem;
  font-weight: 600;
  padding-bottom: 0.5rem;
`;

export const SearchButtonContainer = styled.div`
  padding-top: 0;
  @media only screen and (min-width: 768px) {
    padding-top: 1.5rem;
  }
`;

export const DatePickerWrapper = styled.div`
  .react-datepicker-wrapper,
  .react-datepicker__input-container input {
    width: 100%;
  }
  margin: 0 0 1rem;
  @media only screen and (min-width: 768px) {
    .react-datepicker-wrapper,
    .react-datepicker__input-container input {
      width: auto;
    }
    margin: 0 1.875rem 0 0;
  }
`;

export const StyledButton = styled(Button)`
  width: 100%;
  @media only screen and (min-width: 768px) {
    width: auto;
  }
`;

export const NoRecordsContainer = styled.div`
  height: 22rem;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #F8F8F8;
  font-size: 1.375rem;
  font-weight: 500;
  @media only screen and (min-width: 768px) {
    height: 24.5rem;
    font-size: 1.5rem;
  }
`;

export const ProductContainer = styled.div`
  display: flex;
  flex-direction: column;
  @media only screen and (min-width: 768px) {
    flex-direction: row;
  }
`;

export const ImageContainer = styled.div`
  border: 0.0625rem solid #e8e8e8;
  padding: 1rem;
  margin: 0 0 1rem;
  @media only screen and (min-width: 768px) {
    margin: 0 0 2rem;
  }
`;

export const StyledImage = styled.img`
  height: 200px;
  width: 100%;
  cursor: pointer;
  @media only screen and (min-width: 768px) {
    height: 100px;
    width: 100px;
  }
`;

export const DetailsContainer = styled.div`
  padding: 0 0 1rem;
  @media only screen and (min-width: 768px) {
    padding: 0 1rem 1rem;
  }
`;

export const StyledProductName = styled.div`
  font-size: 1.5rem;
  font-weight: bold;
`;

export const StyledProductDate = styled.div`
  font-style: italic;
  padding-top: 0.3rem;
`;

export const ProductLikeContainer = styled.div`
  display: flex;
  padding-top: 0.5rem;
`;

export const StyledProductLike = styled.img`
  cursor: pointer;
`;

export const StyledLikeCount = styled.span`
  padding-left: 0.3rem;
`;
