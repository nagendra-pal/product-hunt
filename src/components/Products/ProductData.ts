import moment from "moment";
import {ProductType} from "./ProductTypes";

const ProductData: Array<ProductType> = [
    {
        "id" : 1,
        "name": "Wireless headphones",
        "image": "https://res.cloudinary.com/dsrkwxxo7/image/upload/v1589218153/samples/products/Wireless-headphones_vxnlz2.jpg",
        "desc": "Lorem ipsum dolor amet post-ironic four dollar toasts selvage meh Lorem ipsum dolor amet post-ironic four dollar toasts selvage meh Lorem ipsum dolor amet post-ironic four dollar toasts selvage meh",
        "likes": 4,
        "comments": [
            {
                "text": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."
            },
            {
                "text": "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat." 
            },
            {
                
                "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur."
                
            },
            {
                
                "text": "Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
            }
        ],
        "created": moment.utc("2020-03-25").toDate()
    },
    {
        "id" : 2,
        "name": "Round white watch with white band",
        "image": "https://res.cloudinary.com/dsrkwxxo7/image/upload/v1589218152/samples/products/White-watch_n3ansu.jpg",
        "desc": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
        "likes": 10,
        "comments": [
            {
                
                "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur."
                
            },
            {
                
                "text": "Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
            }
        ],
        "created": moment.utc("2020-04-10").toDate()
    },
    {
        "id" : 3,
        "name": "Unpaired maroon plimsoll on top of yellow textile",
        "image": "https://res.cloudinary.com/dsrkwxxo7/image/upload/v1589218152/samples/products/Maroon-plimsoll_i8lhlr.jpg",
        "desc": "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
        "likes": 5,
        "comments": [
            {
                "text": "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat." 
            }
        ],
        "created": moment.utc("2020-04-15").toDate()
    },
    {
        "id" : 4,
        "name": "Beige wooden bar stool",
        "image": "https://res.cloudinary.com/dsrkwxxo7/image/upload/v1589218152/samples/products/Wooden-stool_bex90w.jpg",
        "desc": "Lorem ipsum dolor amet post-ironic four dollar toasts selvage meh Lorem ipsum dolor amet post-ironic four dollar toasts selvage meh Lorem ipsum dolor amet post-ironic four dollar toasts selvage meh",
        "likes": 20,
        "comments": [
            {
                "text": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."
            },
            {
                "text": "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat." 
            },
            {
                
                "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur."
                
            },
            {
                
                "text": "Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
            },
            {
                "text": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."
            },
            {
                "text": "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat." 
            },
            {
                
                "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur."
                
            },
            {
                
                "text": "Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
            }
        ],
        "created": moment.utc("2020-04-20").toDate()
    },
    {
        "id" : 5,
        "name": "Black sunglasses",
        "image": "https://res.cloudinary.com/dsrkwxxo7/image/upload/v1589218152/samples/products/Black-sunglasses_mffok6.jpg",
        "desc": "Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
        "likes": 7,
        "comments": [
            {
                
                "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur."
                
            },
            {
                
                "text": "Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
            }
        ],
        "created": moment.utc("2020-04-25").toDate()
    }
  ];

  //const ProductData: Array<ProductType> = []; // For no records test

  export default ProductData;