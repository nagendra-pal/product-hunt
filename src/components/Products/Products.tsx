import React, { useState } from "react";
import { Button } from "reactstrap";
import DatePicker from "react-datepicker";
import moment from "moment";
import {Body, Title, SearchContainer, DateSearchLabel, SearchButtonContainer, DatePickerWrapper, StyledButton} from "./styles/Products";
import ProductList from "./ProductList";
import {ProductType, DateType} from "./ProductTypes";
import ProductData from "./ProductData";

const Products = (): JSX.Element => {
  const [newDates, setNewDates] = useState<DateType>({
    startDate: null,
    endDate: null,
  });

  const [productList, setProductList] = useState<Array<ProductType>>(ProductData);

  const filterAfterStartDate = (arr: Array<ProductType>): Array<ProductType> => {
    const startDay = moment(newDates.startDate).utc();
    return arr.filter(
      (product) => startDay.diff(moment(product.created).utc(), "hours") <= 0
    );
  };

  const filterBeforeEndDate = (arr: Array<ProductType>): Array<ProductType> => {
    const endDate = moment(newDates.endDate).utc();
    return arr.filter(
      (product) => endDate.diff(moment(product.created).utc(), "days") >= 0
    );
  };

  const filterAfterStartAndBeforeEnd = (arr: Array<ProductType>): Array<ProductType> => {
    return filterBeforeEndDate(filterAfterStartDate(arr));
  };

  const searchHandle = () => {
    const { startDate, endDate } = newDates;

    // if no dates selected - show all results
    if (!startDate && !endDate) setProductList(ProductData);

    // if only end date picked - show only results that were received before end date
    if (!startDate && endDate) {
      setProductList(filterBeforeEndDate(ProductData));
    }
    // if only start date picked - show only results that were received after start day
    if (startDate && !endDate) {
      setProductList(filterAfterStartDate(ProductData));
    }
    // show results that were received after start date and before end date
    if (startDate && endDate) {
      setProductList(filterAfterStartAndBeforeEnd(ProductData));
    }
  };

  return (
    <Body>
      <Title>Product Hunt</Title>
      <SearchContainer>
        <div>
          <DateSearchLabel>From</DateSearchLabel>
          <DatePickerWrapper>
            <DatePicker
              selected={newDates.startDate}
              onChange={(date) => {
                const updatedDates = { ...newDates, startDate: date };
                setNewDates(updatedDates);
              }}
              maxDate={newDates.endDate}
              isClearable={!!newDates.startDate}
              //dateFormat={widthData.moreSm ? 'standard' : 'short'}
              initialDate={newDates.startDate}
              />
            </DatePickerWrapper>
        </div>
        <div>
          <DateSearchLabel>To</DateSearchLabel>
          <DatePickerWrapper>
          <DatePicker
            selected={newDates.endDate}
            onChange={(date) => {
              const updatedDates = { ...newDates, endDate: date };
              setNewDates(updatedDates);
            }}
            minDate={newDates.startDate}
            isClearable={!!newDates.endDate}
            //dateFormat={widthData.moreSm ? 'standard' : 'short'}
            initialDate={newDates.endDate}
          />
          </DatePickerWrapper>
        </div>
        <SearchButtonContainer>
          <StyledButton color="primary" onClick={searchHandle}>
            Search
          </StyledButton>
        </SearchButtonContainer>
      </SearchContainer>
      <ProductList productList={productList} />
    </Body>
  );
};

export default Products;
