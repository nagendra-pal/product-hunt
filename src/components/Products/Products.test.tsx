import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Products from './Products';

describe('Product component', () => {
   it('should renders without crashing', () => {
      const output = shallow(<Products />);
      expect(shallowToJson(output)).toMatchSnapshot();
    });
});

// test('should render Product component correctly', () => {
//    const output = shallow(<Products />);
//    expect(shallowToJson(output)).toMatchSnapshot();
// });

