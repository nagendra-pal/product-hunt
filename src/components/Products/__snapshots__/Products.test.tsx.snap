// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product component should renders without crashing 1`] = `
<styled.div>
  <styled.div>
    Product Hunt
  </styled.div>
  <styled.div>
    <div>
      <styled.div>
        From
      </styled.div>
      <styled.div>
        <o
          allowSameDay={false}
          customTimeInput={null}
          dateFormat="MM/dd/yyyy"
          dateFormatCalendar="LLLL yyyy"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="scroll"
          enableTabLoop={true}
          excludeScrollbar={true}
          initialDate={null}
          inlineFocusSelectedMonth={false}
          isClearable={false}
          maxDate={null}
          monthsShown={1}
          nextMonthButtonLabel="Next Month"
          nextYearButtonLabel="Next Year"
          onBlur={[Function]}
          onCalendarClose={[Function]}
          onCalendarOpen={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onInputError={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          preventOpenOnFocus={false}
          previousMonthButtonLabel="Previous Month"
          previousYearButtonLabel="Previous Year"
          readOnly={false}
          renderDayContents={[Function]}
          selected={null}
          shouldCloseOnSelect={true}
          showFullMonthYearPicker={false}
          showMonthYearPicker={false}
          showPopperArrow={true}
          showPreviousMonths={false}
          showQuarterYearPicker={false}
          showTimeInput={false}
          showTimeSelect={false}
          showYearPicker={false}
          strictParsing={false}
          timeCaption="Time"
          timeInputLabel="Time"
          timeIntervals={30}
          withPortal={false}
        />
      </styled.div>
    </div>
    <div>
      <styled.div>
        To
      </styled.div>
      <styled.div>
        <o
          allowSameDay={false}
          customTimeInput={null}
          dateFormat="MM/dd/yyyy"
          dateFormatCalendar="LLLL yyyy"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="scroll"
          enableTabLoop={true}
          excludeScrollbar={true}
          initialDate={null}
          inlineFocusSelectedMonth={false}
          isClearable={false}
          minDate={null}
          monthsShown={1}
          nextMonthButtonLabel="Next Month"
          nextYearButtonLabel="Next Year"
          onBlur={[Function]}
          onCalendarClose={[Function]}
          onCalendarOpen={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onInputError={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          preventOpenOnFocus={false}
          previousMonthButtonLabel="Previous Month"
          previousYearButtonLabel="Previous Year"
          readOnly={false}
          renderDayContents={[Function]}
          selected={null}
          shouldCloseOnSelect={true}
          showFullMonthYearPicker={false}
          showMonthYearPicker={false}
          showPopperArrow={true}
          showPreviousMonths={false}
          showQuarterYearPicker={false}
          showTimeInput={false}
          showTimeSelect={false}
          showYearPicker={false}
          strictParsing={false}
          timeCaption="Time"
          timeInputLabel="Time"
          timeIntervals={30}
          withPortal={false}
        />
      </styled.div>
    </div>
    <styled.div>
      <Styled(Button)
        color="primary"
        onClick={[Function]}
      >
        Search
      </Styled(Button)>
    </styled.div>
  </styled.div>
  <ProductList
    productList={
      Array [
        Object {
          "comments": Array [
            Object {
              "text": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
            },
            Object {
              "text": "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
            },
            Object {
              "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
            },
            Object {
              "text": "Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            },
          ],
          "created": 2020-03-25T00:00:00.000Z,
          "desc": "Lorem ipsum dolor amet post-ironic four dollar toasts selvage meh Lorem ipsum dolor amet post-ironic four dollar toasts selvage meh Lorem ipsum dolor amet post-ironic four dollar toasts selvage meh",
          "id": 1,
          "image": "https://res.cloudinary.com/dsrkwxxo7/image/upload/v1589218153/samples/products/Wireless-headphones_vxnlz2.jpg",
          "likes": 4,
          "name": "Wireless headphones",
        },
        Object {
          "comments": Array [
            Object {
              "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
            },
            Object {
              "text": "Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            },
          ],
          "created": 2020-04-10T00:00:00.000Z,
          "desc": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
          "id": 2,
          "image": "https://res.cloudinary.com/dsrkwxxo7/image/upload/v1589218152/samples/products/White-watch_n3ansu.jpg",
          "likes": 10,
          "name": "Round white watch with white band",
        },
        Object {
          "comments": Array [
            Object {
              "text": "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
            },
          ],
          "created": 2020-04-15T00:00:00.000Z,
          "desc": "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
          "id": 3,
          "image": "https://res.cloudinary.com/dsrkwxxo7/image/upload/v1589218152/samples/products/Maroon-plimsoll_i8lhlr.jpg",
          "likes": 5,
          "name": "Unpaired maroon plimsoll on top of yellow textile",
        },
        Object {
          "comments": Array [
            Object {
              "text": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
            },
            Object {
              "text": "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
            },
            Object {
              "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
            },
            Object {
              "text": "Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            },
            Object {
              "text": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
            },
            Object {
              "text": "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
            },
            Object {
              "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
            },
            Object {
              "text": "Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            },
          ],
          "created": 2020-04-20T00:00:00.000Z,
          "desc": "Lorem ipsum dolor amet post-ironic four dollar toasts selvage meh Lorem ipsum dolor amet post-ironic four dollar toasts selvage meh Lorem ipsum dolor amet post-ironic four dollar toasts selvage meh",
          "id": 4,
          "image": "https://res.cloudinary.com/dsrkwxxo7/image/upload/v1589218152/samples/products/Wooden-stool_bex90w.jpg",
          "likes": 20,
          "name": "Beige wooden bar stool",
        },
        Object {
          "comments": Array [
            Object {
              "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
            },
            Object {
              "text": "Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            },
          ],
          "created": 2020-04-25T00:00:00.000Z,
          "desc": "Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
          "id": 5,
          "image": "https://res.cloudinary.com/dsrkwxxo7/image/upload/v1589218152/samples/products/Black-sunglasses_mffok6.jpg",
          "likes": 7,
          "name": "Black sunglasses",
        },
      ]
    }
  />
</styled.div>
`;
